cache/manager.go

```diff
@@ -49,6 +50,9 @@ type Accessor interface {
 	Get(ctx context.Context, id string, opts ...RefOption) (ImmutableRef, error)
 
 	New(ctx context.Context, parent ImmutableRef, s session.Group, opts ...RefOption) (MutableRef, error)
+	// GetMutable returns a mutable reference to the cache record with the given ID if it is possible to do so. If the
+	// provided cache record ID points to a committed but not finalized record, than a new cache record will be returned
+	// that is a copy of the previous but with a new ID (in order to invalidate any references to the previous ID).
```

Comment:
It will only change ID if there used to be an immutable ref open w/ the previous ID.

So, if you have a ref open as mutable, release it, and then call `GetMutable`, the ID will not change.

If, however, you have a ref open as mutable, commit it to an immutable ref, release it and then call `GetMutable`, the ID will change in order to invalidate anyone who stored the ID of the previous immutable ref.

Reply:
It still seems very hacky and unexpected. I'm not sure if you are rewriting indexes then or how otherwise the mutable reuse could work at all. The point of reuse is that 1) mutable x is created, id is stored 2) turned immutable y 3) x is not available 4) immutable released 5) x can be accessed again by the ID and modified 6) turned immutable z, as data changed and pointers to y and z should not be equal.

Reply:
> I'm not sure if you are rewriting indexes then or how otherwise the mutable reuse could work at all

There's a new util function in the metadata code for copying the metadata/indexes from one id to another, which is what's used to create the new mutable ref: https://github.com/moby/buildkit/pull/2223/files#diff-ae44b7efd0878d4ad4013864764f0046c5c75f5f7dfe08718ad3bd32cabf9caeR401

> The point of reuse is that 1) mutable x is created, id is stored 2) turned immutable y 3) x is not available 4) immutable released 5) x can be accessed again by the ID and modified 6) turned immutable z, as data changed and pointers to y and z should not be equal.

Makes sense, with the code in this PR the important constraints of that behavior are retained even if the semantics are slightly altered. Using your example, the modified behavior in this PR is:
1. record x is created with a mutable ref pointing to it, id is stored
2. mutable ref to x is made immutable
3. mutable ref to x can't be opened
4. immutable ref to x released
5. mutable ref to x is requested, ID turns to y
6. mutable ref to y turned immutable, as data changed pointers to x and y are not equal

While it's obviously subjective, the reason I prefer this modified behavior over the previous is that I think it both simplifies the mental model and the actual code.

Here's the model this PR is aiming for:
* There are cache records, identified by an ID, which represent a single piece of snapshot (and related blob) state.
* Mutable and Immutable refs are sort of like "smart pointers" to cache records that enforce safe access patterns and implement reference counting.
* IDs will change when a mutable ref is reopened for a cache record which previously had an immutable ref pointing to it

Here's the model before this PR as I understand it:
* There are cache records, identified by an ID, which represent either a single piece of snapshot+blob state or can themselves be pointers to other "equivalent" records (via the `equal*` fields) while still retaining their own ID.
* Mutable and Immutable Refs are smart pointers like above but the rules for access patterns and reference counting now span not only the cache records they point to but also any cache records those cache records might themselves point to via `equal*`.
* IDs will change when a mutable ref is committed into an immutable ref

The fact that IDs change when a committed record is reopened as mutable instead of when mutable is committed to immutable seems like a fine trade given the other simplifications. It also worked out-of-the-box with most of the existing callers anyways because they just used search indexes instead of directly storing mutable IDs.

In terms of the actual code, the `equal*` fields seemed to create convolution all over the place in the cache code (see the previous interaction between them in the release methods of [immutableRef](https://github.com/moby/buildkit/blob/124126e7517d5dfdf5695a7eb8e357b9743d1e61/cache/refs.go#L757) and [mutableRef](https://github.com/moby/buildkit/blob/124126e7517d5dfdf5695a7eb8e357b9743d1e61/cache/refs.go#L933), the need for stuff like the [doubleRef field](https://github.com/moby/buildkit/blob/124126e7517d5dfdf5695a7eb8e357b9743d1e61/cache/manager.go#L883), etc.). The new model trades that convolution for what amounts mostly to isolated boilerplate (such as constructing a new ref in GetMutable).

Let me know if this makes more sense or if instead you consider it important that we retain the exact same behavior as before, in which case I'll just go ahead and close this PR as I'm not sure how to simplify under those constraints and don't want to drag this out forever.

Reply:
I tend to think this weird ID behavior, and duplicating indexes on getting a record adds more confusion than that local `doubleRef` variable. I wouldn't be even sure indexes can be moved like this, certainly was not designed with this capability in mind. I guess the implementation with a single cacherecord that maintains previous behavior and updates its immutable ID is still possible but with the current interface behavior and added lines I don't think it makes things much simpler.

Reply:
Okay, I disagree pretty strongly that the behavior is weird and more confusing but I've got other PRs to send out so I'll just close this for now. I hope that in the future we can figure out some way to simplify this so the release methods and other parts of the code impacted by this are more tractable, but for now I'll just try to avoid modifying any of the equal* stuff to the extent possible.